KIP-290 Prefixed ACLs

big-andy-coates opened the pull request:
[KIP-290](https://cwiki.apache.org/confluence/display/KAFKA/KIP-290%3A+Support+for+wildcard+suffixed+ACLs) /  [KAFKA-6841](https://issues.apache.org/jira/browse/KAFKA-6841)

This PR supersedes https://github.com/apache/kafka/pull/5079. (As I lack rights to push to that PR).

Still to do (can be done in this or another PR):
- [x] Doc updates, including upgrade.html
- [x] AclCommand updates a.k.a. kafka-acls.sh
- [x] Remove internal use of deprecated constructors, (want to make sure the tests pass using legacy first)
- [x] Performance improvements in the SimpleAclAuthorizer.getMatchingAcls, (which currently visits every resource).

The following outstanding items have been moved to additional Jiras or deemed unnecessary. See https://github.com/apache/kafka/pull/5117#issuecomment-395066985.
- Remove duplicate Java `Resource` class.
- Remove duplicate Scala `Resource` class.
- Remove  duplicate Scala `ResourceNameType` class and just Java one.
- Use single `/kafka-acl-changes` path and store ResourceNameType a JSON Value.
- Potentially replacing `Resource` in `AclBinding` with `ResourceFilter` or a `ResourceMatcher` 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

big-andy-coates committed:
KAFKA-6841: Piyush changes

big-andy-coates committed:
Add testing and Admin client changes.

big-andy-coates commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -198,31 +202,54 @@ class SimpleAclAuthorizer extends Authorizer with Logging {
+    inReadLock(lock) {
+      aclCache
+        .filterKeys(aclResource => AclUtils.matchResource(aclResource.toJava, filter))
+        .flatMap(_._2.acls)
```
@junrao, if you can suggest an alternative to this `flatmap(_._2.acls)` I'm happy to plug it in, but my Scala's a little rusty.

junrao replied:
You can just do .flatMap{ case(resource, versionedAcls) => versionedAcls.acls} .

big-andy-coates committed:
upgrade.html update.

big-andy-coates commented:
test this please

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/CreateAclsRequest.java:
```diff
@@ -51,9 +52,19 @@
 
     /**
-     * The version number is bumped to indicate that on quota violation brokers send out responses before throttling.
+     * V1 sees a new `RESOURCE_NAME_TYPE` that controls how the resource name is interpreted.
```
We want to preserve the comment on quota violation.

big-andy-coates replied:
done.

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/DeleteAclsRequest.java:
```diff
@@ -52,9 +53,19 @@
 
     /**
-     * The version number is bumped to indicate that on quota violation brokers send out responses before throttling.
+     * V1 sees a new `RESOURCE_NAME_TYPE_FILTER` that controls how the filter handles different resource name types.
```
We want to preserve the comment on quota violation.

big-andy-coates replied:
done.

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/DeleteAclsResponse.java:
```diff
@@ -67,12 +84,20 @@
 
     /**
-     * The version number is bumped to indicate that on quota violation brokers send out responses before throttling.
+     * V1 sees a new `RESOURCE_NAME_TYPE` field added to MATCHING_ACL_V1, that describes how the resource name is interpreted.
```
We want to preserve the comment on quota violation.

big-andy-coates replied:
done.

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeAclsRequest.java:
```diff
@@ -44,9 +45,18 @@
 
     /**
-     * The version number is bumped to indicate that on quota violation brokers send out responses before throttling.
+     * V1 sees a new `RESOURCE_NAME_TYPE_FILTER` that controls how the filter handles different resource name types.
```
We want to preserve the comment on quota violation.

big-andy-coates replied:
done

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeAclsResponse.java:
```diff
@@ -57,16 +58,37 @@
 
     /**
-     * The version number is bumped to indicate that on quota violation brokers send out responses before throttling.
+     * V1 sees a new `RESOURCE_NAME_TYPE` field added to DESCRIBE_ACLS_RESOURCE_V1, that describes how the resource name is interpreted.
```
We want to preserve the comment on quota violation.

big-andy-coates replied:
done.

junrao commented on clients/src/main/java/org/apache/kafka/common/resource/Resource.java:
```diff
@@ -22,14 +22,15 @@
 
 /**
- * Represents a cluster resource with a tuple of (type, name).
+ * Represents a cluster resource with a tuple of (type, name, nameType).
```
Should we remove the Resource in o.a.k.c.request?

big-andy-coates replied:
IMHO, no.

Conceptually, it doesn't make any sense for a _resource_ to support literal or prefixed names. A _resource_ should be an actual cluster resource, with a full name.

The two `Resource` classes are used in two different areas of the API: topic config and Acls.  For one area were being forced to add this `ResourceTypeName` to the `Resource` class. But in the other are the name type field is not required.

I would say that we leave these two classes as separate for now and look to merge them in KIP-50. This way, users wanting to make use of the topic config API don't need to worry about this new `ResourceNameType` field. This is only required if dealing with Acls, and only until we implement KIP-50, which should also make the field redundant.

cmccabe replied:
To be fair, `org.apache.kafka.common.requests.Resource` is not actually a part of the public AdminClient API for getting and setting configurations.  The public class for that is `org.apache.kafka.common.config.ConfigResource`.  I would argue that we should just get rid of `o.a.k.c.r.Resource`, since it just duplicates `ConfigResource`.  We just end up copying things to and from `ConfigResource` in `alterConfigs` and the other functions.  And it's an internal private class anyway, so we can always re-create it if we need to (we won't)

(I would also argue that ConfigResource should never have been separate from `org.apache.kafka.common.resource.Resource`, but that ship has sailed since ConfigResource is part of the stable API now).

cmccabe replied:
By the way, I think we should remove `o.a.k.c.r.Resource` in a separate follow-on change.  It's not really related to KIP-290

junrao commented on core/src/main/scala/kafka/security/auth/Resource.scala:
```diff
@@ -16,31 +16,44 @@
+ * @param resourceNameType non-null type of resource name: literal, prefixed, etc.
  */
-case class Resource(resourceType: ResourceType, name: String) {
+case class Resource(resourceType: ResourceType, name: String, resourceNameType: ResourceNameType) {
```
Should ClusterResource be created using this constructor?

big-andy-coates replied:
Yes, eventually. I wanted to submit an initial version of the code, that still uses the deprecated constructor everywhere, to show that test cases still pass with these constructors.  I've a 'todo' in the PRs description to switch to the new constructors. I'll do that today.

junrao commented on core/src/main/scala/kafka/security/auth/ResourceNameType.scala:
```diff
@@ -0,0 +1,48 @@
+package kafka.security.auth
+
+import kafka.common.{BaseEnum, KafkaException}
+import kafka.utils.ZkUtils
```
unused import

big-andy-coates replied:
done.

junrao commented on core/src/main/scala/kafka/security/auth/ResourceNameType.scala:
```diff
@@ -0,0 +1,48 @@
+
+  def values: Seq[ResourceNameType] = List(Literal, Prefixed)
+
+  def fromJava(operation: JResourceNameType): ResourceNameType = fromString(operation.toString.replaceAll("_", ""))
```
operation => jResourceNameType ? Also, why are we replacing _?

big-andy-coates replied:
Looks like Piyush cut & paste this from `ResourceType`. It's not strictly needed here. Ill drop.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -27,11 +26,13 @@ import kafka.security.auth.SimpleAclAuthorizer.VersionedAcls
 import kafka.utils._
-import kafka.zk.{AclChangeNotificationSequenceZNode, AclChangeNotificationZNode, KafkaZkClient}
+import kafka.zk.{AclChangeNotificationSequenceZNode, KafkaZkClient, ZkAclStore}
+import org.apache.kafka.common.resource.{ResourceFilter, Resource => JResource, ResourceNameType => JResourceNameType}
```
unused import Resource.

big-andy-coates replied:
done.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -27,11 +26,13 @@ import kafka.security.auth.SimpleAclAuthorizer.VersionedAcls
+import org.apache.kafka.common.utils.{AclUtils, SecurityUtils, Time}
 
 import scala.collection.JavaConverters._
+import scala.collection.immutable.Stream.Empty
```
unused import

big-andy-coates replied:
done.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -198,31 +202,54 @@ class SimpleAclAuthorizer extends Authorizer with Logging {
+
+    inReadLock(lock) {
+      aclCache
+        .filterKeys(aclResource => AclUtils.matchResource(aclResource.toJava, filter))
```
This seems less efficient than before since it scans all stored acls.

big-andy-coates replied:
Yep, there is a todo in the description of the PR to cover this - I have a plan ;)

junrao commented on core/src/main/scala/kafka/zk/KafkaZkClient.scala:
```diff
@@ -1056,20 +1062,22 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
   /**
-   * Gets the resource names for a give resource type
-   * @param resourceType
+   * Gets the resource names, for which ACLs are stored, for a give resource type and name type
```
a give => a given

big-andy-coates replied:
done

junrao commented on core/src/main/scala/kafka/zk/ZkData.scala:
```diff
@@ -26,9 +26,9 @@ import kafka.cluster.{Broker, EndPoint}
+import kafka.security.auth.{Acl, Literal, Prefixed, Resource, ResourceNameType, ResourceType}
 import kafka.server.{ConfigType, DelegationTokenManager}
-import kafka.utils.Json
+import kafka.utils.{Json, ZkUtils}
```
unused import ZkUtils

big-andy-coates replied:
done.

junrao commented on core/src/main/scala/kafka/zk/ZkData.scala:
```diff
@@ -40,7 +40,7 @@ import org.apache.zookeeper.data.{ACL, Stat}
 import scala.collection.JavaConverters._
 import scala.collection.mutable.ArrayBuffer
-import scala.collection.{Seq, breakOut}
+import scala.collection.{GenTraversableOnce, Seq, breakOut}
```
unused import

big-andy-coates replied:
done.

junrao commented on clients/src/test/java/org/apache/kafka/common/utils/AclUtilsTest.java:
```diff
@@ -0,0 +1,202 @@
+    public void shouldMatchPrefixedIfNamePrefixedAnyFilterTypeIsAny() {
+        assertTrue(AclUtils.matchResource(
+            new Resource(TOPIC, "Name", PREFIXED),
+            new ResourceFilter(TOPIC, "Name-something", ResourceNameType.ANY)
```
This makes sense from authorize()'s perspective. However, for AdminClient, does this make sense?

big-andy-coates replied:
Yes, see below.

junrao commented on clients/src/test/java/org/apache/kafka/common/utils/AclUtilsTest.java:
```diff
@@ -0,0 +1,202 @@
+    public void shouldMatchLiteralWildcardIfFilterHasNameTypeOfAny() {
+        assertTrue(AclUtils.matchResource(
+            new Resource(TOPIC, "*", LITERAL),
+            new ResourceFilter(TOPIC, "Name", ResourceNameType.ANY)
```
This makes sense from authorize()'s perspective. However, for AdminClient, does this make sense?

big-andy-coates replied:
Yes, I think so.  This works well to allow users to _discover_ all the ACLs affecting a resource, be they literal, wildcard or prefixed.

# Old functionality:
Each call retrieves ACLs stored in one path in ZK:
`adminClient.describeAcls(... TOPIC, "foobar" ...)` -> would return only ACLs from '/kafka-acls/foobar' path
`adminClient.describeAcls(... TOPIC, "*" ...)` -> would return only ACLs from '/kafka-acls/*' path

# New functionality:
## Using legacy constructors:
Legacy constructors default resourceNameType to Literal and maintains existing contract:
`adminClient.describeAcls(... TOPIC, "foobar" ...)` -> still returns only ACLs from '/kafka-acls/foobar' path
`adminClient.describeAcls(... TOPIC, "*" ...)` -> still returns only ACLs from '/kafka-acls/*' path

## Using constructors with explicit resourceNameType:
The user needs to _know_, up front, which prefixed resource paths exist to be able to query them:
`adminClient.describeAcls(... TOPIC, "foobar", Literal ...)`-> will return only ACLs from '/kafka-acls/foobar' path
`adminClient.describeAcls(... TOPIC, "*", Literal ...)` -> will return only ACLs from '/kafka-acls/*' path
`adminClient.describeAcls(... TOPIC, "foo", Prefixed ...)` -> will return only ACLs from '/kafka-prefixed-acls/foo' path

## Using 'ANY' resource name type:
This allows user to _discover_ all the ACLs affecting a specific resource:
`adminClient.describeAcls(... TOPIC, "foobar", Any ...)` -> will return all ACLs affecting topic "foobar", including any prefixed and wildcard ACLs.

The return value from `describeAcls` will contain the resourceNameType field for each ACL, so the user can determine if it is literal or prefixed.

Acls on prefixed resource paths are _never_ returned to earlier clients. Nor can older clients delete ACLs on prefixed resource paths.

junrao replied:
Sounds good. It would be useful to add the following comment when we make the changes in adminClient.

This allows user to discover all the ACLs affecting a specific resource:
adminClient.describeAcls(... TOPIC, "foobar", Any ...) -> will return all ACLs affecting topic "foobar", including any prefixed and wildcard ACLs.

big-andy-coates replied:
Yer, I already had that in a local java doc change. Pushed.

junrao commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -745,7 +745,7 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
 
       // now authorize the user for the internal topic and verify that we can subscribe
-      addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Read)), Resource(Topic,
+      addAndVerifyAcls(Set(Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Read)), new Resource(Topic,
```
Should we use the new constructor for Resource?

big-andy-coates replied:
Yes, as mentioned above, I'll do this all in one pass now that the tests are all passing.

junrao commented on core/src/test/scala/integration/kafka/api/SaslSslAdminClientIntegrationTest.scala:
```diff
@@ -88,12 +89,16 @@ class SaslSslAdminClientIntegrationTest extends AdminClientIntegrationTest with
     closeSasl()
   }
 
+  val anyAcl = new AclBinding(new Resource(ResourceType.TOPIC, "*"),
```
Should we use the new Resource constructor? Ditto below.

big-andy-coates replied:
Yes, as mentioned above, I'll do this all in one pass now that the tests are all passing.

junrao commented on core/src/test/scala/unit/kafka/security/auth/SimpleAclAuthorizerTest.scala:
```diff
@@ -419,6 +433,107 @@ class SimpleAclAuthorizerTest extends ZooKeeperTestHarness {
+    assertEquals(2, simpleAclAuthorizer.getAcls(principal).size)
+
+    val acl2 = new Acl(Acl.WildCardPrincipal, Allow, WildCardHost, Write)
+    simpleAclAuthorizer.addAcls(Set[Acl](acl1), new Resource(Group, "groupA"))
```
Should we use the new Resource constructor?

big-andy-coates replied:
Yes, as mentioned above, I'll do this all in one pass now that the tests are all passing.

junrao reviewed (commented):
@big-andy-coates : Thanks for the patch. A few comments below. Also, (1) I agree that we can fix AclBinding in the adminClient to just include AclFilter in a followup patch in this release. We can fix SimpleAuthorizer as part of KIP-50. (2) We can probably just consolidate the acl changes on /kafka-acl-changes. During down grade, we will load the full acls from the resources on startup. So, as long as subsequent acl changes are made through the old format. It should be fine.

big-andy-coates committed:
- review comments
- AclCommand (kafka-acls.sh) now supported prefixed
- Documentation for kafka-acls.has updated for prefixed.
- Optimised SimpleAclAuthorizer.getMatchingAcls call.

big-andy-coates committed:
- don't use deprecated functions internally

big-andy-coates commented:
@junrao latest version pushed. Includes:
- changes inline with your review comments
- AclCommand a.k.a. kafka-acls.sh now supports '--resource-name-type' flag, with options 'literal', 'prefixed', or 'any'. 
- Documentation updated for kafka-acls.sh
- Some optimisation within `SimpleAclAuthroizer.getMatchingAcls`, though TBH, I'm still not happy with it. I'm not sure `TreeMap` is the way to go here - any thoughts welcome!
- Move away from use of deprecated constructors.

Are you OK with merging this and picking up follow on changes in separate PRs? Still outstanding:
- Remove duplicate Resource class - which I don't think we should do
 - Use single /kafka-acl-changes path and store ResourceNameType a JSON Value.
 - Potentially replacing Resource in AclBinding with ResourceFilter or a ResourceMatcher

big-andy-coates committed:
Merge branch 'trunk' into prefixed-acls

# Conflicts:
#	core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala
#	docs/upgrade.html

big-andy-coates committed:
self review.

big-andy-coates commented:
@junrao Please note that Piyush may wish to resubmit these changes under his own name once we're all happy with the code. So hold of merging until Piyush has made his wisher clear, (or until the deadline looms!)

piyushvijay commented on clients/src/main/java/org/apache/kafka/common/protocol/CommonFields.java:
```diff
@@ -45,6 +46,8 @@
     public static final Field.Str RESOURCE_NAME = new Field.Str("resource_name", "The resource name");
     public static final Field.NullableStr RESOURCE_NAME_FILTER = new Field.NullableStr("resource_name", "The resource name filter");
+    public static final Field.Int8 RESOURCE_NAME_TYPE = new Field.Int8("resource_name_type", "The resource name type", AclUtils.LITERAL.code());
+    public static final Field.Int8 RESOURCE_NAME_TYPE_FILTER = new Field.Int8("resource_name_type", "The resource name type filter", AclUtils.LITERAL.code());
```
s/resource_name_type/resource_name_type_filter?

big-andy-coates replied:
done.

big-andy-coates commented:
test this please

big-andy-coates commented:
@junrao, @piyushvijay is happy for us to merge from this PR, but could you please call out his contribution in the commit message?

big-andy-coates committed:
review changes

big-andy-coates commented:
test this please. (unrelated test failure).

junrao commented on docs/security.html:
```diff
@@ -1143,14 +1156,26 @@ <h4><a id="security_authz_examples" href="#security_authz_examples">Examples</a>
+            <pre class="brush: bash;">bin/kafka-acls.sh --authorizer-properties zookeeper.connect=localhost:2181 --add --allow-principal User:Peter --allow-host 198.51.200.1 --producer --topic *</pre>
+            You can add acls on resources matching a certain prefix, e.g. suppose you want to add an acl "Principal User:Jane is allowed to produce to any Topic whose name is prefixed with 'Test-' from any host".
+            You can do that by executing the CLI with following options:
+            <pre class="brush: bash;">bin/kafka-acls.sh --authorizer-properties zookeeper.connect=localhost:2181 --add --allow-principal User:Jane --producer --topic Test- --resource-name-type Prefixed</pre>
```
Test- => Test-topic? Ditto below in a few other places.

big-andy-coates replied:
The 'Test-' is deliberate. This command is adding a prefixed ACL, i.e. one matching any Topic whose name starts with 'Test-', so it would match 'Test-topic'

cmccabe commented on clients/src/main/java/org/apache/kafka/common/requests/CreateAclsRequest.java:
```diff
@@ -51,9 +52,20 @@
     /**
-     * The version number is bumped to indicate that on quota violation brokers send out responses before throttling.
+     * V1 sees a new `RESOURCE_NAME_TYPE` that controls how the resource name is interpreted and version
+     * was bumped to indicate that, on quota violation, brokers send out responses before throttling.
```
I think this could be phrased a little bit better.  Something like "Version 1 adds RESOURCE_NAME_TYPE.  Also, when the quota is violated, brokers will respond to a version 1 or later request before throttling."

big-andy-coates replied:
I thought so to, but didn't want to change other peoples wording! Consider it done.

cmccabe commented on clients/src/main/java/org/apache/kafka/common/requests/CreateAclsRequest.java:
```diff
@@ -51,9 +52,20 @@
+                    HOST,
+                    OPERATION,
+                    PERMISSION_TYPE))));
 
```
I can't figure out how to add a comment to the correct line using github's terrible user interface.  But on line 110 where you have the code:

```
        @Override
        public CreateAclsRequest build(short version) {
            return new CreateAclsRequest(version, creations);
        }
```
You must add verification that the version you are using supports the ACLs you are trying to send.  If someone tries to set a prefix ACL when using CREATE_ACLS_REQUEST_V0, the code must throw `UnsupportedVersionException`.  In fact, anything besides `ResourceNameType.LITERAL` should trigger a UVE there.

big-andy-coates replied:
Consider it done.

cmccabe commented on clients/src/main/java/org/apache/kafka/common/requests/CreateAclsResponse.java:
```diff
@@ -43,7 +43,8 @@
     /**
-     * The version number is bumped to indicate that on quota violation brokers send out responses before throttling.
+     * The version number is bumped to indicate that, on quota violation, brokers send out responses before throttling
+     * and request has additional resource_name_type field.
```
It's not necessary to duplicate the comment about CreateAclsRequest inside CreateAclsResponse.  The response doesn't have any additional fields.  I don't think there's anything to document here beyond the original comment about quota stuff.

big-andy-coates replied:
done.

cmccabe commented:
Can you add a unit or integration test of `UnsupportedVersionException` being thrown when someone tries to create a prefix ACL on a v0 server?  This is important for compatibility.

> Potentially replacing Resource in AclBinding with ResourceFilter or a ResourceMatcher

That would be a breaking API change, which would mean that current users of the API would need to be rewritten.

I agree that `Resource` is a bit of misnomer now (Maybe `Resources` or `ResourcesIdentifier` would have been a better class name)  However, ResourceFilter is not the correct class to use here.  The "Filter" classes are for use in deleteAcls.

A quick example might help to explain.  When I remove `AclBindingFilter(ResourceFilter(name=*,...))`, I don't want to remove all ACLs, even though the star (wildcard) matches everything.  I want to remove the wildcard ACL, if it exists.  The filter specifically checks the name, and verifies that it is "*".

The filter can have fields set to null, in which case they match everything.  So if I want to remove all ACLs, can remove `AclBindingFilter(ResourceFilter(name=null,...))`  We need to support both removing all ACLs, and removing the star ACL specifically.  They are not the same thing.  Same thing with prefix ACLs, etc.  Removing the prefix ACL foo-bar should not remove all ACLs that start with foo-bar

Overall, I think we should leave the class names as-is.  I don't think there's a lot to be gained by changing it now, and it would be a big inconvenience for users to break compatibility.

Although it doesn't need to be done in this PR, I do think we should remove the internal classes which just duplicate the Resource class.  I'm thinking of `o.a.k.c.r.Resource`, as well as the Scala version of Resource. I guess you could argue that the internal classes could express an additional type constraint in some cases (and I see you made that argument here).  But I think any slight benefit there is dwarfed by the  annoying boilerplate code to convert back and forth.  We have to validate the Resource data we get from the caller anyway, so we will already have runtime type constraints (you can't create something with type UNKNOWN, etc.)

cmccabe commented on clients/src/main/java/org/apache/kafka/common/utils/AclUtils.java:
```diff
@@ -0,0 +1,79 @@
+import org.apache.kafka.common.resource.ResourceNameType;
+import org.apache.kafka.common.resource.ResourceType;
+
+public final class AclUtils {
```
I don't understand why we need a separate class and a static method for this.  The only caller is `ResourceFilter#matches`.  That's where this logic was before, and I think that's where it still belongs.

big-andy-coates replied:
Agreed, this is a hang over of previous changes.

cmccabe commented on clients/src/main/java/org/apache/kafka/common/utils/AclUtils.java:
```diff
@@ -0,0 +1,79 @@
+            case LITERAL:
+                return aclPath.name().equals(filter.name()) || aclPath.name().equals(WILDCARD_RESOURCE);
+
+            case PREFIXED:
```
This is wrong.  AclFilter objects should match exactly or not at all.  Trying to remove a wildcard AclBinding should not remove every AclBinding.  Trying to remove a prefix AclBinding should not remove every AclBinding with that prefix.

big-andy-coates replied:
This KIP / PR has to see the semantics of matching a filter to a resource change.  Please see https://github.com/apache/kafka/pull/5117#discussion_r192668669

cmccabe replied:
Sorry, but this is wrong.  There are two "semantics of matching".
1. The semantics of an AclBindingFilter matching an AclBinding
2. The semantics of an AclBinding matching an ACL

Prefix comparisons are only relevant to #2, not to #1.

When I try to remove an AclBinding for a prefix ACL beginning with "foo", I don't want to remove ACLs that start with foobar, foo2, etc.  When I remove a wildcard ACL, I don't want to remove all ACLs.  And so forth.

big-andy-coates replied:
This code is purely responsible for matching a Resource to a ResourceFilter. There are no ACLs involved. So as far as I can tell there is only one matching semantic here, which is the matching of a ResourceFilter and a Resource.

If trying to remove an AclBinding for a prefixed topic beginning with 'foo' then the filter would be new ResourceFilter(Topic, "foo", Prefixed), which would NOT match a `Resource(Topic, "foobar", Prefixed)` or indeed `Resource(Topic, "foobar", Literal)`.  There are unit tests covering this.

big-andy-coates replied:
Have you maybe missed the:

```
if (nameType().equals(other.nameType())) {
    return other.name().equals(name());
}
```

statement that means any match between a filter and resource of the same ResourceNameType requires an exact match on the name?

cmccabe replied:
To summarize our offline discussion, the proposal is to override ResourceNameType.ANY in a somewhat odd, but potentially useful way.  Let's pick up this discussion in a follow on PR.

cmccabe commented on core/src/main/scala/kafka/security/auth/ResourceNameType.scala:
```diff
@@ -0,0 +1,49 @@
+/*
```
Can't we just use the Java class here rather than duplicating it?

The only reason the other classes are duplicated is that the Scala class came first, and later the Java classes were added for the clients.

big-andy-coates replied:
Sure, I'll clean up in a follow on PR. (It's past midnight here!)

cmccabe replied:
Thanks, @big-andy-coates.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -298,28 +335,44 @@ class SimpleAclAuthorizer extends Authorizer with Logging {
     }
   }
 
+  private object ResourceOrdering extends Ordering[Resource] {
```
Could we add a comment that we want to sort in reverse resource name order?

big-andy-coates replied:
I'm not a fan of such comments, that can easily end up out of sync with the code. But I can add one.

junrao commented on core/src/main/scala/kafka/admin/AclCommand.scala:
```diff
@@ -232,23 +258,25 @@ object AclCommand extends Logging {
+  private def getResourceFilter(opts: AclCommandOptions, dieIfNoResourceFound: Boolean = true): Set[ResourceFilter] = {
+    val resourceNameType: JResourceNameType = opts.options.valueOf(opts.resourceNameType)
+
+    var resources = Set.empty[ResourceFilter]
```
resources => resourceFilters?

big-andy-coates replied:
done.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -27,9 +26,10 @@ import kafka.security.auth.SimpleAclAuthorizer.VersionedAcls
+import org.apache.kafka.common.resource.{ResourceFilter, ResourceNameType => JResourceNameType}
 import org.apache.kafka.common.security.auth.KafkaPrincipal
-import org.apache.kafka.common.utils.{SecurityUtils, Time}
+import org.apache.kafka.common.utils.{AclUtils, SecurityUtils, Time}
```
unused import AclUtils

big-andy-coates replied:
done.

junrao commented on core/src/test/scala/unit/kafka/admin/AclCommandTest.scala:
```diff
@@ -82,45 +83,73 @@ class AclCommandTest extends ZooKeeperTestHarness with Logging {
+
+  @Test
+  def testAclsOnPrefixedResources(): Unit = {
+    val cmd = Array("--allow-principal", principal.toString, "--producer", "--topic", "Test-", "--resource-name-type", "Prefixed")
```
Test- => Test-topic ?

big-andy-coates replied:
Again, this is intentional as its adding a ACLs on a prefixed resource name.

junrao reviewed (commented):
@big-andy-coates : Thanks for the patch. Just a few more minor comments below.

piyushvijay commented:
Hi @big-andy-coates, 

As discussed offline, there is a standard way to add co-authors: https://blog.github.com/2018-01-29-commit-together-with-co-authors/ (cc: @cmccabe @junrao). Probably add me before merging.

Thanks

big-andy-coates commented:
@cmccabe 

> Can you add a unit or integration test of UnsupportedVersionException being thrown when someone tries to create a prefix ACL on a v0 server? This is important for compatibility.

Yep, done. 

>> Potentially replacing Resource in AclBinding with ResourceFilter or a ResourceMatcher

> That would be a breaking API change, which would mean that current users of the API would need to be rewritten.

> I agree that Resource is a bit of misnomer now (Maybe Resources or ResourcesIdentifier would have been a better class name) However, ResourceFilter is not the correct class to use here. The "Filter" classes are for use in deleteAcls.

a) I need to remind myself of the reason why this was necessary!
b) I'm sure we can make backwards compatible, deprecated constructors etc if the change is warranted.

I'll come back to you in a separate PR.

> A quick example might help to explain. When I remove AclBindingFilter(ResourceFilter(name=*,...)), I don't want to remove all ACLs, even though the star (wildcard) matches everything. I want to remove the wildcard ACL, if it exists. The filter specifically checks the name, and verifies that it is "*".

> The filter can have fields set to null, in which case they match everything. So if I want to remove all ACLs, can remove AclBindingFilter(ResourceFilter(name=null,...)) We need to support both removing all ACLs, and removing the star ACL specifically. They are not the same thing. Same thing with prefix ACLs, etc. Removing the prefix ACL foo-bar should not remove all ACLs that start with foo-bar

The changes in this PR are completely backwards compatible, both in terms of working with older clients and in terms of maintaining the existing contracts with new clients. Additionally, they extend the semantics by adding `ResourceNameType` to the potential filter parameters.  This is necessary to allow users to discover prefixed ACLs.  Full details https://github.com/apache/kafka/pull/5117#discussion_r192668669 and in the updated KIP.

> Overall, I think we should leave the class names as-is. I don't think there's a lot to be gained by changing it now, and it would be a big inconvenience for users to break compatibility.

You may be right. I'll remind myself of the reasons and, if appropriate, come back with a case and a PR that we can discuss.

> Although it doesn't need to be done in this PR, I do think we should remove the internal classes which just duplicate the Resource class. I'm thinking of o.a.k.c.r.Resource, as well as the Scala version of Resource. I guess you could argue that the internal classes could express an additional type constraint in some cases (and I see you made that argument here). But I think any slight benefit there is dwarfed by the annoying boilerplate code to convert back and forth. We have to validate the Resource data we get from the caller anyway, so we will already have runtime type constraints (you can't create something with type UNKNOWN, etc.)

As I've mentioned in comments to Jun, I don't think we should tackle removing these under this KIP. o.a.k.c.r.Resource is used exclusively for dealing with topic config, i.e. o.a.k.c.r.Resource only ever refers to a concrete topic resource, and never the wildcard topic or a prefixed topic. `ResourceNameType` makes zero sense in this context.  They system can't support users requesting to add config to 'Resource(Topic, 'foo-", Prefixed)'. 

I think this work is better done under KIP-50. Which, BTW, I'm happy to pick up.  In this KIP we can split the `Resource` type into something representing a concrete resource, (i.e. without `ResourceNameType`), and something else like `ResourceMatcher` and use these appropriately in all of the AdminClient, AclCommand, topic config and authorizer areas.

Are you pushing for these duplicate classes to be dropped now because the next release is a major version bump?

big-andy-coates committed:
review changesreview changes

big-andy-coates commented:
@junrao  @cmccabe really appreciate your feed back - changes pushed and some comments pushed back on.

big-andy-coates commented:
@cmccabe, just seen your comment: 

> To be fair, org.apache.kafka.common.requests.Resource is not actually a part of the public AdminClient API for getting and setting configurations. The public class for that is org.apache.kafka.common.config.ConfigResource. I would argue that we should just get rid of o.a.k.c.r.Resource, since it just duplicates ConfigResource. We just end up copying things to and from ConfigResource in alterConfigs and the other functions. And it's an internal private class anyway, so we can always re-create it if we need to (we won't)

> (I would also argue that ConfigResource should never have been separate from org.apache.kafka.common.resource.Resource, but that ship has sailed since ConfigResource is part of the stable API now).

That makes more sense - so we don't replace the use of  org.apache.kafka.common.requests.Resource with  org.apache.kafka.common.resource.Resource, but with org.apache.kafka.common.config.ConfigResource.

big-andy-coates committed:
self review

big-andy-coates committed:
Add additional test.

big-andy-coates committed:
Add additional test.

big-andy-coates commented:
test this please

ijuma committed:
Merge remote-tracking branch 'apache-github/trunk' into kip-290-acl-prefix

* apache-github/trunk:
  KAFKA-6726: Fine Grained ACL for CreateTopics (KIP-277) (#4795)
  KAFKA-5588: Remove deprecated --new-consumer tools option (#5097)
  MINOR: Fix for the location of the trogdor.sh executable file in the documentation. (#5040)
  KAFKA-6997: Exclude test-sources.jar when $INCLUDE_TEST_JARS is FALSE
  MINOR: docs should point to latest version (#5132)
  KAFKA-6981: Move the error handling configuration properties into the ConnectorConfig and SinkConnectorConfig classes (KIP-298)
  [KAFKA-6730] Simplify State Store Recovery (#5013)
  MINOR: Rename package `internal` to `internals` for consistency (#5137)
  KAFKA-6704: InvalidStateStoreException from IQ when StreamThread closes store (#4801)
  MINOR: Add missing configs for resilience settings
  MINOR: Add regression tests for KTable mapValues and filter (#5134)
  KAFKA-6750: Add listener name to authentication context (KIP-282) (#4829)
  KAFKA-3665: Enable TLS hostname verification by default (KIP-294) (#4956)
  KAFKA-6938: Add documentation for accessing Headers on Kafka Streams Processor API (#5128)
  KAFKA-6813: return to double-counting for count topology names (#5075)
  KAFKA-5919; Adding checks on "version" field for tools using it
  MINOR: Remove deprecated KafkaStreams constructors in docs (#5118)

ijuma commented:
I merged this PR with trunk and resolved conflicts.

ijuma commented:
I synced with Jun offline and he is happy for this PR to be merged and for unresolved issues to be addressed in follow-up PRs. I will merge this once the tests pass and will set Piyush and Andy as a co-authors. @big-andy-coates, can you please file one or more JIRAs with "Fix version" 2.0.0 for the remaining work?

big-andy-coates commented:
OK, have created the following Jira:

* [KAFKA-7005: Remove duplicate Java Resource class](https://issues.apache.org/jira/browse/KAFKA-7005)
* [KAFKA-7006: Remove duplicate Scala ResourceNameType class](https://issues.apache.org/jira/browse/KAFKA-7006)
* [KAFKA-7007: All ACL changes should use single /kafka-acl-changes path](https://issues.apache.org/jira/browse/KAFKA-7007)
* [KAFKA-7008: Consider replacing the Resource field in AclBinding with a ResourceFilter or ResourceMatcher](https://issues.apache.org/jira/browse/KAFKA-7008)

I've not created a Jira to  "Remove duplicate Scala Resource class", which @cmccabe suggested, as I've realised that this would require changes to the `Authorizer` interface, which would cause a lot of issue for companies with their own implementations. Instead, I propose this scala class is deprecated as part of KIP-50. Shout if any one disagrees with this.

ijuma commented:
Thanks @big-andy-coates
